### PR TITLE
ddl: check allocator's existence before allocate/rebase

### LINKED ISSUE #18931: create table with table option 'auto_increment' reports nil pointer dereference
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. turn off the config `alter-primary-key`.
2. ```sql
   create table t (a int primary key) AUTO_INCREMENT = 10;
    ```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

```console
Query OK, 0 rows affected
Time: 0.012s
```

### 3. What did you see instead (Required)

```console
(1105, 'runtime error: invalid memory address or nil pointer dereference')
```

### 4. Affected version (Required)

master
```console
Release Version: v4.0.0-beta.2-870-g2a8b96845
Edition: Community
Git Commit Hash: 2a8b968453520e4fcf9d6ff46c9f23b4ad23feee
Git Branch: HEAD
UTC Build Time: 2020-08-03 03:14:54
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### 5. Root Cause Analysis

Before allocate/rebase values, a nil check for allocators is missed.

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #18931 

Problem Summary: 

After #18326, `rowid` allocator is not necessarily existed in every table. Before allocate/rebase values, a nil check for allocators is missed.

### What is changed and how it works?

What's Changed and how it works: add a nil check for allocators before using them.

### Related changes

- Need to cherry-pick to the release branch together with https://github.com/pingcap/tidb/pull/18326.

### Check List 

Tests 

- Unit test
- Integration test

Side effects: N/A

### Release note 

- No release note

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1970,22 +1972,11 @@ func checkCharsetAndCollation(cs string, co string) error {
...
-	// The operation of the minus 1 to make sure that the current value doesn't be used,
-	// the next Alloc operation will get this value.
-	// Its behavior is consistent with MySQL.
-	if tp == autoid.RowIDAllocType {
-		if err = tb.RebaseAutoID(nil, tbInfo.AutoIncID-1, false, tp); err != nil {
-			return errors.Trace(err)
-		}
-	} else {
-		if err = tb.RebaseAutoID(nil, tbInfo.AutoRandID-1, false, tp); err != nil {
+	if alloc := allocs.Get(tp); alloc != nil {
```
Comment (djshow832) [inline ddl/ddl_api.go:1977]: So the manual rebase will be just ignored when `alloc==nil`?

Comment (tangenta) [reply ddl/ddl_api.go:1977]: Yes. This is consistent with MySQL.

File: ddl/db_integration_test.go [hunk 2]
```diff
@@ -2328,3 +2328,17 @@ func (s *testIntegrationSuite4) TestAlterIndexVisibility(c *C) {
...
+	defer config.RestoreFunc()()
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.AlterPrimaryKey = false
+	})
+	tk.MustExec("drop database if exists test_create_table_option_auto_inc;")
+	tk.MustExec("create database test_create_table_option_auto_inc;")
+	tk.MustExec("use test_create_table_option_auto_inc;")
+
+	tk.MustExec("create table t (a bigint primary key) auto_increment = 10;")
+	tk.MustExec("alter table t auto_increment = 10;")
```
Comment (djshow832) [inline ddl/db_integration_test.go:2343]: Try this:
```sql
create table t (a bigint primary key) auto_increment = 10;
alter table t add column b bigint unique key auto_increment;
insert t value(1, 0);
select b from t;
alter table t auto_increment=10;
insert t value(2, 0);
select b from t;
```
The result seems right, but it deserves a trial.

Comment (tangenta) [reply ddl/db_integration_test.go:2343]:
```console
tidb> alter table t add column b bigint unique key auto_increment;
(8200, "unsupported add column 'b' constraint UNIQUE KEY when altering 'test.t'")
```

Comment (tangenta) [reply ddl/db_integration_test.go:2343]: In MySQL, the result is 
```
+----+
| b  |
+----+
| 1  |
| 10 |
+----+
```

Comment (djshow832) [reply ddl/db_integration_test.go:2343]: So TiDB doesn't support add a unique key at all?

File: go.mod [hunk 3]
```diff
@@ -61,6 +61,7 @@ require (
 	golang.org/x/tools v0.0.0-20200527183253-8e7acdbce89d
 	google.golang.org/grpc v1.26.0
 	gopkg.in/natefinch/lumberjack.v2 v2.0.0
+	honnef.co/go/tools v0.0.1-2020.1.5 // indirect
```
Comment (crazycs520) [inline go.mod:64]: ?

Comment (tangenta) [reply go.mod]: ?

**Pull request merged**